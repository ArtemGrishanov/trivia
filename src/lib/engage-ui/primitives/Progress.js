import React from 'react'
import DataSchema from '../../schema'
import RemixWrapper from '../RemixWrapper'
import TextEditor from '../bricks/TextEditor'
import { CompletionIcon } from '../icons'
import '../style/rmx-progress.css'

/**
 *
 * @param {object} props
 * @param {string} props.color
 * @param {string} props.completionColor
 * @param {boolean} props.isFirst
 * @param {boolean} props.isLast
 * @param {boolean} props.isCompleted
 */
function ProgressArrowItem({
    color = '#D8D8D8',
    completionColor = '#69B1FC',
    isFirst = false,
    isLast = false,
    isCompleted = false,
}) {
    const backgroundColor = isCompleted ? completionColor : color
    let className = 'progress-variant-2__arrow'

    if (isFirst) {
        className += ' progress-variant-2__arrow_first'
    } else if (isLast) {
        className += ' progress-variant-2__arrow_last'
    }

    return (
        <div className={className} style={{ backgroundColor }}>
            {isCompleted && <CompletionIcon className="progress-variant-2__completed-icon" />}
        </div>
    )
}

/**
 *
 * @param {object} props
 * @param {number} props.radius
 * @param {string} props.color
 * @param {string} props.completionColor
 * @param {boolean} props.isCompleted
 */
function ProgressDotItem({ radius = 6, color = '#C4C4C4', completionColor = '#2990FB', isCompleted = false }) {
    const backgroundColor = isCompleted ? completionColor : color
    return <div className="progress-variant-3__dot" style={{ width: radius, height: radius, backgroundColor }} />
}

class Progress extends React.Component {
    constructor(props) {
        super(props)
        this.state = {}
    }

    getTemplate = () => {
        const {
            variant,
            step,
            max,
            completionBackground,
            background,
            fontSize,
            color,
            borderRadius,
            borderWidth,
            borderColor,
            dotSize,
            text,
            doubleClicked,
            id,
        } = this.props
        switch (variant) {
            case 'variant1': {
                const percent = +(step / max).toFixed(2) * 100
                const resultText = text
                    .replace(/<([^>]+)>/gi, '')
                    .replace(/\d+%/gi, '')
                    .trim()
                return (
                    <div className="progress-variant-1">
                        <div style={{ marginBottom: 8, textAlign: 'left', fontSize, color }}>
                            <TextEditor parentId={id} readOnly={!doubleClicked} text={resultText + ` ${percent}%`} />
                        </div>
                        <div
                            style={{
                                width: '100%',
                                boxSizing: 'border-box',
                                borderStyle: 'solid',
                                borderRadius,
                                borderWidth,
                                borderColor,
                                background,
                            }}
                        >
                            <div
                                style={{
                                    width: `${percent}%`,
                                    backgroundColor: completionBackground,
                                    height: 8,
                                    borderRadius,
                                }}
                            ></div>
                        </div>
                    </div>
                )
            }
            case 'variant2': {
                return (
                    <div className="progress-variant-2">
                        {new Array(max).fill('').map((_, index) => {
                            const isFirst = index === 0
                            const isLast = index === max - 1
                            const isCompleted = index < step
                            return (
                                <ProgressArrowItem
                                    key={index}
                                    color={background}
                                    completionColor={completionBackground}
                                    isFirst={isFirst}
                                    isLast={isLast}
                                    isCompleted={isCompleted}
                                />
                            )
                        })}
                    </div>
                )
            }
            case 'variant3': {
                return (
                    <div className="progress-variant-3">
                        {new Array(max).fill('').map((_, index) => {
                            const isCompleted = index < step
                            return (
                                <ProgressDotItem
                                    key={index}
                                    color={background}
                                    radius={dotSize}
                                    completionColor={completionBackground}
                                    isCompleted={isCompleted}
                                />
                            )
                        })}
                    </div>
                )
            }
            case 'variant0':
            default: {
                const st = {
                    fontSize,
                    color,
                }
                const text = this.props.step + '/' + this.props.max
                return (
                    <p className="rmx-text" style={st}>
                        {text}
                    </p>
                )
            }
        }
    }

    render() {
        let template = this.getTemplate()
        return template
    }
}

/**
 * Props schema
 * Which props could be edited and how (types, range and other rules)
 */
export const Schema = new DataSchema({
    text: {
        type: 'string',
        minLength: 1,
        maxLength: 4096,
        default: 'Progress:',
    },
    variant: {
        type: 'string',
        enum: ['variant0', 'variant1', 'variant2', 'variant3'],
        default: 'variant0',
    },
    width: {
        type: 'number',
        min: 1,
        max: 1000,
        default: 32,
    },
    height: {
        type: 'number',
        min: 1,
        max: 1000,
        default: 25,
    },
    step: {
        type: 'number',
        min: 1,
        max: 1000,
        default: 1,
    },
    max: {
        type: 'number',
        min: 1,
        max: 1000,
        default: 10,
    },
    fontSize: {
        type: 'number',
        min: 8,
        max: 80,
        default: 14,
    },
    color: {
        type: 'string',
        default: '#3C3C3C',
    },
    background: {
        type: 'string',
        default: '#D8D8D8',
    },
    completionBackground: {
        type: 'string',
        default: '#2990FB',
    },
    borderRadius: {
        type: 'number',
        min: 0,
        max: 100,
        default: 16,
    },
    borderWidth: {
        type: 'number',
        min: 0,
        max: 100,
        default: 0,
    },
    borderColor: {
        type: 'string',
        default: '',
    },
    dotSize: {
        type: 'number',
        min: 1,
        max: 100,
        default: 6,
    },
})

export default RemixWrapper(Progress, Schema, 'Progress')
