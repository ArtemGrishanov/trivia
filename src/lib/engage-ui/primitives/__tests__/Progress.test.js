import React from 'react'
import { shallow } from 'enzyme'

import { Progress, ProgressArrowItem, ProgressDotItem } from './../Progress'

describe('<Progress />', () => {
    let wrapper
    let props

    beforeAll(() => {
        props = {
            step: 2,
            max: 5,
            fontSize: 14,
            color: 'green',
        }
    })

    it('should match <Progress /> snapshot as variant0', () => {
        wrapper = shallow(<Progress {...props} variant="variant0" />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should match <Progress /> snapshot as variant1', () => {
        wrapper = shallow(
            <Progress
                {...props}
                text="Progress:"
                variant="variant1"
                progressText="Progress:"
                background="#f3f3f3"
                completionBackground="#c5e5e5"
                borderColor="#000"
                borderRadius={5}
                borderWidth={1}
                width={32}
            />,
        )
        expect(wrapper).toMatchSnapshot()
    })

    it('should match <Progress /> snapshot as variant2', () => {
        wrapper = shallow(<Progress {...props} variant="variant2" color="grey" completionColor="blue" />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should match <Progress /> snapshot as variant3', () => {
        wrapper = shallow(<Progress {...props} variant="variant3" color="grey" radius={10} completionColor="blue" />)
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<ProgressArrowItem />', () => {
    let wrapper
    let props

    beforeAll(() => {
        props = {
            height: 16,
            width: 125,
            color: 'grey',
            completionColor: 'blue',
            isFirst: false,
            isLast: false,
            isCompleted: false,
        }
    })

    it('should match snapshot', () => {
        wrapper = shallow(<ProgressArrowItem {...props} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should match <ProgressArrowItem /> snapshot when it is first', () => {
        wrapper = shallow(<ProgressArrowItem {...props} isFirst={true} />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should match <ProgressArrowItem /> snapshot when it is last', () => {
        wrapper = shallow(<ProgressArrowItem {...props} isLast={true} />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should match <ProgressArrowItem /> snapshot when it is completed', () => {
        wrapper = shallow(<ProgressArrowItem {...props} isCompleted={true} />)
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<ProgressDotItem />', () => {
    let wrapper
    let props

    beforeAll(() => {
        props = {
            radius: 10,
            color: 'grey',
            completionColor: 'blue',
            isCompleted: false,
        }
    })

    it('should match snapshot', () => {
        wrapper = shallow(<ProgressDotItem {...props} />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should match <ProgressDotItem /> snapshot when it is completed', () => {
        wrapper = shallow(<ProgressDotItem {...props} isCompleted={true} />)
        expect(wrapper).toMatchSnapshot()
    })
})
