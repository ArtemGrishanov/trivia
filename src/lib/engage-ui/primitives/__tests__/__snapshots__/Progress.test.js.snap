// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Progress /> should match <Progress /> snapshot as variant0 1`] = `
<p
  className="rmx-text"
  style={
    Object {
      "color": "green",
      "fontSize": 14,
    }
  }
>
  2/5
</p>
`;

exports[`<Progress /> should match <Progress /> snapshot as variant1 1`] = `
<div
  className="progress-variant-1"
>
  <div
    style={
      Object {
        "color": "green",
        "fontSize": 14,
        "marginBottom": 8,
        "textAlign": "left",
      }
    }
  >
    <TextEditor
      readOnly={true}
      text="Progress: 40%"
    />
  </div>
  <div
    style={
      Object {
        "background": "#f3f3f3",
        "borderColor": "#000",
        "borderRadius": 5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "boxSizing": "border-box",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#c5e5e5",
          "borderRadius": 5,
          "height": 8,
          "width": "40%",
        }
      }
    />
  </div>
</div>
`;

exports[`<Progress /> should match <Progress /> snapshot as variant2 1`] = `
<div
  className="progress-variant-2"
>
  <ProgressArrowItem
    isCompleted={true}
    isFirst={true}
    isLast={false}
    key="0"
  />
  <ProgressArrowItem
    isCompleted={true}
    isFirst={false}
    isLast={false}
    key="1"
  />
  <ProgressArrowItem
    isCompleted={false}
    isFirst={false}
    isLast={false}
    key="2"
  />
  <ProgressArrowItem
    isCompleted={false}
    isFirst={false}
    isLast={false}
    key="3"
  />
  <ProgressArrowItem
    isCompleted={false}
    isFirst={false}
    isLast={true}
    key="4"
  />
</div>
`;

exports[`<Progress /> should match <Progress /> snapshot as variant3 1`] = `
<div
  className="progress-variant-3"
>
  <ProgressDotItem
    isCompleted={true}
    key="0"
  />
  <ProgressDotItem
    isCompleted={true}
    key="1"
  />
  <ProgressDotItem
    isCompleted={false}
    key="2"
  />
  <ProgressDotItem
    isCompleted={false}
    key="3"
  />
  <ProgressDotItem
    isCompleted={false}
    key="4"
  />
</div>
`;

exports[`<ProgressArrowItem /> should match <ProgressArrowItem /> snapshot when it is completed 1`] = `
<div
  className="progress-variant-2__arrow"
  style={
    Object {
      "backgroundColor": "blue",
    }
  }
>
  <CompletionIcon
    className="progress-variant-2__completed-icon"
  />
</div>
`;

exports[`<ProgressArrowItem /> should match <ProgressArrowItem /> snapshot when it is first 1`] = `
<div
  className="progress-variant-2__arrow progress-variant-2__arrow_first"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
/>
`;

exports[`<ProgressArrowItem /> should match <ProgressArrowItem /> snapshot when it is last 1`] = `
<div
  className="progress-variant-2__arrow progress-variant-2__arrow_last"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
/>
`;

exports[`<ProgressArrowItem /> should match snapshot 1`] = `
<div
  className="progress-variant-2__arrow"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
/>
`;

exports[`<ProgressDotItem /> should match <ProgressDotItem /> snapshot when it is completed 1`] = `
<div
  className="progress-variant-3__dot"
  style={
    Object {
      "backgroundColor": "blue",
      "height": 10,
      "width": 10,
    }
  }
/>
`;

exports[`<ProgressDotItem /> should match snapshot 1`] = `
<div
  className="progress-variant-3__dot"
  style={
    Object {
      "backgroundColor": "grey",
      "height": 10,
      "width": 10,
    }
  }
/>
`;
